### PR TITLE
Optimize graphical display with selective rendering

### PR DESCRIPTION
**Pull request by thinkyhead**
The draw loop wastes a lot of cycles because it re-calculates the whole display 2 or more times per screen render. This PR employs two strategies to reduce rendering time:

- Add bounds-checking so that only items within the current page being rendered are drawn.
- For XYZ —overlapping both halves of the screen— the value strings are cached.
- Use a frame with normal text instead of a solid box with inverse text.
- Reduce the delay between bytes sent over SPI by 5-6µs.

Added up, these optimizations save at least ~19ms (380,000 cycles) on screens with 2 stripes.

Additionally:
- Use `uint8_t` for u8g coordinates. (u8g also supplies a `u8g_uint_t` type)
- Use `const` for function arguments and locals, where appropriate.

I look forward to your test results!

### CODE REVIEW

**Comment by Sebastianv650** [comment]
This one is better than #5279: It saves 2ms (82.5 to 80.5ms) on my printer.

**Comment by thinkyhead** [comment]
2ms is significant! That's 32,000 cycles.

@olikraus suggests we can do even better with u8g2. But it is still a work-in-progress…

Please read https://github.com/olikraus/u8glib/issues/447#issuecomment-262707911

**Comment by thinkyhead** [comment]
Also @yhfudev take note! U8g2 apparently has some helpful features that may make your work easier. Specifically:

> - "UTF-8 support, Chinese fonts are already included, limit of 256 glyphs per font removed."

Proper UTF-8 support means we don't have to do so much remapping.

> - Very fast, LiquidCrystal like char based API (U8x8) if graphics output is not required.

Only the Info screen needs graphics. Other screens can take advantage of this optimization.

https://github.com/olikraus/u8glib/issues/447#issuecomment-262707911

**Comment by AnHardt** [comment]
Very good!

Todays tests:

Todays test of the GLCD enhacements

RCBugFix Status screen
2 Stripes 77ms
4 Stripes 86ms
8 Stripes 110ms

Selective rendering
2 Stripes 75ms (-2ms)
4 Stripes 76ms (-10ms)
8 Stripes 85ms (-25ms)

Selective + unroll
2 Stripes 78ms (+1ms)

Selective + unroll + decrease delay
2 Stripes 76ms (-1ms)

Selective + decrease delay
2 Stripes 73ms (-4ms)

Detailed timings in the commits comments. (https://github.com/AnHardt/Marlin/pull/67)

Conclusion:
Selective rendering is a small win for the two stripe ST7920 but makes a huge difference for displays with more stripes. Go for it.
Unrolling the loops seems to make the updates a tiny bit slower. Don't merge.
Decreasing the 10µs-delay to 5Ms speeds up ~2ms per complete update. Today i could not go below 5µs. If we should do that - with the original 10 µseconds and a big fat warning - decrease on your own risk.

**Comment by ghost** [comment]
I also tested.

| Selective + 10μs delay | Selective + 5μs delay | Selective + 4μs delay | Selective + 3μs delay |
| :-: | :-: | :-: | :-: |
| 2 Stripes | 2 Stripes | 2 Stripes | 2 Stripes |
| 74 ~ 75ms | 73 ~ 74ms | 72 ~ 73ms | 72 ~ 73ms |
| ~~Displaying is broken~~ It's displayed properly  | It's displayed properly ~~, but bit noisy when encoder is rotated~~ | It's displayed properly ~~, but bit noisy when encoder is rotated~~ | Displaying is broken |
|  | | | dscn1749 |

~~A branch that used for testing: https://github.com/esenapaj/Marlin/tree/glcd-testes~~

~~And, I created a branch that it doesn't include benchmark and 5μs patch and tested. https://github.com/esenapaj/Marlin/tree/glcd-testes2~~
~~Displaying is broken.~~ It's displayed properly.
~~Video clip:~~

**Comment by AnHardt** [comment]
@thinkyhead 
Would you pleas check https://github.com/AnHardt/Marlin/pull/67#issuecomment-262859051.

**Comment by AnHardt** [comment]
@esenapaj 
Please reset the display (shut down power completely) and retest __Selective__ with the normal 10µs delay. Simply rebooting the Arduino sometimes does not work when the delays previously have been too short..

**Comment by ghost** [comment]
@AnHardt 
Sorry and thanks for the advice, now I've re-tested and confirmed that both this PR and this PR + 5μs patch works without any problems.

**Comment by ghost** [comment]
@AnHardt
I've tested your "frame" solution.

Selective + 5μs delay + "frame"
2 Stripes
58 ~59ms (-19ms)
It's displayed properly.

Selective + 4μs delay + "frame"
2 Stripes
58 ~59ms (-19ms)
It's displayed properly.

Selective + 3μs delay + "frame"
2 Stripes
57 ~58ms (-20ms)
Displaying is broken.

~~A branch that used for testing: https://github.com/esenapaj/Marlin/tree/glcd-testes3~~

**Comment by AnHardt** [comment]
Tomorrow i'll make some experiments with 'hollow' extruder and sd-card symbols.

**Comment by thinkyhead** [comment]
These are really encouraging results! A savings of 19ms frees up 304,000 CPU cycles. Wow.

Can you also test with the options `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` to make sure the display still works correctly, including menus and edit items?

Also, try the display with and without `LCD_SCREEN_ROT_180`. Oliver suggested that the `page_y0` and `page_y1` values might be reversed in that case.

I've added an option `DOGM_SPI_DELAY_US` for tuning that 10µs delay. It's in `Configuration_adv.h`.

Drawing the frame instead of a solid box is also very smart. As another alternative I was also thinking of trying pre-clipping the solid box. That is, instead of drawing the whole box, starting at 30 with a height of `INFO_FONT_HEIGHT + 2`, make sure to only draw the portion that's within the current stripe.

Or… place that part of the display at the very top of the screen instead of in the middle, so it's only ever drawn once (including the XYZ strings).

**Comment by ghost** [comment]
I got a compilation error.
<details>
<summary>error message</summary>

```
In file included from \AppData\Local\Temp\arduino_build_683885\sketch\ultralcd_impl_DOGM.h:46:0,

                 from \AppData\Local\Temp\arduino_build_683885\sketch\ultralcd.cpp:56:

\AppData\Local\Temp\arduino_build_683885\sketch\ultralcd_st7920_u8glib_rrd.h: In function 'void marlin_SPI_delay()':

ultralcd_st7920_u8glib_rrd.h:112: error: '_delay_system_ticks_sub' was not declared in this scope

       _delay_system_ticks_sub((CYCLES_PER_MICROSECOND) * (DOGM_SPI_DELAY_US));

                                                                             ^
```
</details>

**Comment by thinkyhead** [comment]
Do you think that `delayMicroseconds` is fine? For some reason the u8g library has that custom delay method, `_delay_system_ticks_sub` but it's maybe not publicly accessible. (I'll try `extern` and see if that fixes it anyway.)

**Comment by ghost** [comment]
Still got a compilation error
<details>
<summary>error message</summary>

```
In file included from \AppData\Local\Temp\arduino_build_498533\sketch\ultralcd_impl_DOGM.h:46:0,

                 from \AppData\Local\Temp\arduino_build_498533\sketch\ultralcd.cpp:56:

\AppData\Local\Temp\arduino_build_498533\sketch\ultralcd_st7920_u8glib_rrd.h: In function 'void marlin_SPI_delay()':

ultralcd_st7920_u8glib_rrd.h:112: error: 'delay_system_ticks' was not declared in this scope

       delay_system_ticks((CYCLES_PER_MICROSECOND) * (DOGM_SPI_DELAY_US));

                                                                        ^
```
</details>
, thus I avoid it tentatively.

| Selective + 5μs delay + "frame" | Selective + 5μs delay + "frame" | Selective + 4μs delay + "frame" | Selective + 3μs delay + "frame" |
| :-: | :-: | :-: | :-: |
| without both `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` |
| without `LCD_SCREEN_ROT_180` | without `LCD_SCREEN_ROT_180` | without `LCD_SCREEN_ROT_180` |
| 2 Stripes | 2 Stripes | 2 Stripes |
| 58 ~ 59ms | 55 ~ 56ms | 54 ~ 55ms | 54 ~ 55ms |
| It's displayed properly | It's displayed properly but adjustment is needed | It's displayed properly but adjustment is needed | Displaying is broken |
| dscn1754 | dscn1751 | dscn1751 | |

| without this PR | Selective + 10μs delay + "frame" | Selective + 10μs delay + "frame" | Selective + 5μs delay + "frame" |
| :-: | :-: | :-: | :-: |
| `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | without both `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` |
| with `LCD_SCREEN_ROT_180` | with `LCD_SCREEN_ROT_180` | with `LCD_SCREEN_ROT_180` | with `LCD_SCREEN_ROT_180` |
| 2 Stripes | 2 Stripes | 2 Stripes | 2 Stripes |
| -  | 45 ~ 46ms | 39 ~ 40ms | 37 ~ 38ms |
| It's displayed properly | It isn't displayed properly | It isn't displayed properly | It isn't displayed properly |
| | dscn1753 | dscn1752 | dscn1752 |

~~A branch that used for testing: https://github.com/esenapaj/Marlin/tree/glcd-testes4~~

**Comment by jbrazio** [comment]
@thinkyhead 
> U8g2 apparently has some helpful features

Last time I checked they were still missing PROGMEM support.

**Comment by thinkyhead** [comment]
> Still got a compilation error

Oops, forgot to push. Should be better now.

**Comment by thinkyhead** [comment]
@esenapaj Helpful results. So, the rotated screen needs to be accounted for. Apparently there's a way to access the page bounds that will work with any rotation…
```cpp
u8g.getU8g()->current_page.y0
u8g.getU8g()->current_page.y1
```
…so I'll try adapting the code to use these.

I can also try to fix the positioning of the smaller font.

Are the glyphs for X Y Z broken with the small font?

**Comment by thinkyhead** [comment]
> they were still missing PROGMEM support

Oliver is being very responsive! He has already submitted this for us:

- olikraus/u8g2#94 Required display support for Marlin

The AVR-specific optimizations that make u8g faster than u8g2 are still extant, however.

**Comment by Blue-Marlin** [comment]
Have not analysed this properly, but this and some similar places may cost some performance.
```
In file included from C:\Users\andre\AppData\Local\Temp\arduino_build_353192\sketch\ultralcd.cpp:56:0:
C:\Users\andre\AppData\Local\Temp\arduino_build_353192\sketch\ultralcd_impl_DOGM.h: In function 'void _draw_heater_status(uint8_t, int8_t)':
C:\Users\andre\AppData\Local\Temp\arduino_build_353192\sketch\ultralcd_impl_DOGM.h:189:77: warning: comparison is always true due to limited range of data type [-Wtype-limits]

 #define PAGE_CONTAINS(ya, yb) (page.page_y0 <= (yb) && ((ya) ? page.page_y1 >= (ya) : true))
                                                                             ^
C:\Users\andre\AppData\Local\Temp\arduino_build_353192\sketch\ultralcd_impl_DOGM.h:329:7: note: in expansion of macro 'PAGE_CONTAINS'
```

**Comment by Blue-Marlin** [comment]
from u8glib
```
  u8g_uint_t page_y0;
  u8g_uint_t page_y1;
...
typedef uint8_t u8g_uint_t;
```

Ah. I see. This occurs only when the first parameter is 0. And 0 is always >= any uint.

**Comment by thinkyhead** [comment]
@Blue-Marlin And unfortunately the compiler doesn't notice that that code never runs (and probably isn't even compiled in) if the `ya` parameter is `0`. It must analyze types earlier than it does code-culling.
```
(ya) ? page.page_y1 >= (ya) : true)
```

**Comment by Blue-Marlin** [comment]
Indeed.
```
// For selective rendering within a Y range
#define PAGE_CONTAINS(ya, yb) (u8g.getU8g()->current_page.y0 <= (yb) && ((ya) ? u8g.getU8g()->current_page.y1 >= (ya) : true))
```
is working with LCD_SCREEN_ROT_180.
picture 11
and can be simplified to
```
#define PAGE_CONTAINS(ya, yb) (u8g.getU8g()->current_page.y0 <= (yb) && u8g.getU8g()->current_page.y1 >= (ya))
```

But note - the screen the photo shows takes ~81ms - while without LCD_SCREEN_ROT_180 it takes ~59ms.
LCD_SCREEN_ROT_180 takes seriously time!

**Comment by Blue-Marlin** [comment]
@thinkyhead 
For the partial expression `page.page_y1 >= (ya)` he is absolutely right.

**Comment by thinkyhead** [comment]
Ok, so I threw in a hidden option that you can enable/disable for testing `XYZ_HOLLOW_FRAME`. And the various positioning values are now also given names, so they are easier to move around.
```cpp
#if ENABLED(USE_SMALL_INFOFONT)
  #define INFO_FONT_HEIGHT 7
#else
  #define INFO_FONT_HEIGHT 8
#endif

#define XYZ_BASELINE (30 + INFO_FONT_HEIGHT)

#define X_LABEL_POS  3
#define X_VALUE_POS 11
#define XYZ_SPACING 40

// Enable to save many cycles by drawing a hollow frame
#define XYZ_HOLLOW_FRAME

#if ENABLED(XYZ_HOLLOW_FRAME)
  #define XYZ_FRAME_TOP 29
  #define XYZ_FRAME_HEIGHT INFO_FONT_HEIGHT + 4
#else
  #define XYZ_FRAME_TOP 30
  #define XYZ_FRAME_HEIGHT INFO_FONT_HEIGHT + 2
#endif
```

**Comment by thinkyhead** [comment]
> and can be simplified

Ironically, I was trying to avoid the compiler even seeing the test of `>= (ya)` when `ya` is given as `0` to avoid the type warning. To eliminate it, I guess we have to replace occurrences of lines like `PAGE_CONTAINS(0, 12)` with `PAGE_UNDER(12)`.

**Comment by thinkyhead** [comment]
So also in this PR I sought to optimize the drawing of menus and SD card lists. Each row in these lists is determined by `DOG_CHAR_HEIGHT`, but that also includes the spacing around the text. The menus also draw a solid box, but the rest of the menu drawing is probably more optimal than the Info Screen. So, is it worth it to get into more detail in the optimization of these screens also? As with the Info Screen, drawing a hollow frame is probably going to save as much as 17ms.

**Comment by Blue-Marlin** [comment]
Or we let it as it is. Now, since i understand, this is not a performance problem, but pure compiler ignorance i'm not in sorrow any more.

**Comment by thinkyhead** [comment]
Wow, I'm surprised the rotated screen takes so much more time to render. Is it possible that something else is going on here? For example, maybe the test of `page.page == 0` is passing every time because of the rotated screen, causing the XYZ strings to be generated more…? Or maybe it's something to do with the rotation of the text. Oliver indicated that rotated screens are rendered differently, so maybe the data has to be copied twice every time.

I've started to look at the code of U8glib and I can already see some room for optimization. Maybe this is one area where we can help U8glib itself to get better too.

The thing is, a VIKI display on a RAMPS is rotated 180° by default. Have we already seen reports that VIKI displays perform worse than others?

File: Marlin/Configuration_adv.h [hunk 1]
```diff
@@ -459,15 +459,19 @@
 
 #endif // SDSUPPORT
 
-// for dogm lcd displays you can choose some additional fonts:
+// Some additional options are available for graphical displays:
 #if ENABLED(DOGLCD)
-  // save 3120 bytes of PROGMEM by commenting out #define USE_BIG_EDIT_FONT
-  // we don't have a big font for Cyrillic, Kana
+  // A bigger font is available for edit items. Costs 3120 bytes of PROGMEM.
+  // Only Western languages are supported. Not available for Cyrillic or Kana.
```
Comment (AnHardt) [inline Marlin/Configuration_adv.h]: + Turkish, Greece, Chinese

**Comment by ghost** [comment]
I've re-tested a updated this PR.

| Selective + 5μs delay | Selective + 5μs delay | Selective + 5μs delay | Selective + 5μs delay | >Are the glyphs for X Y Z broken with the small font? |
| :-: | :-: | :-: | :-: | :-: |
| without both `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | without both `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | |
| without `LCD_SCREEN_ROT_180` | with `LCD_SCREEN_ROT_180` | without `LCD_SCREEN_ROT_180` | with `LCD_SCREEN_ROT_180` | |
| 2 Stripes | 2 Stripes | 2 Stripes | 2 Stripes | |
| 58 ~ 59ms | 80 ~ 81ms | 56 ~ 57ms | 78 ~ 79ms | |
| It's displayed properly but adjustment is needed | It's displayed properly but adjustment is needed | It's displayed properly but adjustment is needed | It's displayed properly but adjustment is needed | It looks like that it's no problem |
dscn1755 1 | dscn1756 1 | dscn1757 1 | dscn1759 1 | dscn1758 1 |

~~A branch that used for testing: https://github.com/esenapaj/Marlin/tree/glcd-testes5~~

**Comment by jbrazio** [comment]
Using a simple u8glib test program with those small OLED screens makes quite noticeable the slower draw speed when it is rotated (you can see it just by looking at it). I would not believe something is wrong with your code Scott.

**Comment by AnHardt** [comment]
@thinkyhead 
Do you remember summer 2015, when some VIKI2 users began to complain about slow displays? VIKI2 is the only display with `LCD_SCREEN_ROT_180` (and 8/8).

**Comment by thinkyhead** [comment]
> I would not believe something is wrong with your code Scott.

That's my default position too! 😉 

> Do you remember summer 2015, when the some VIKI2 users began to complain

Yes, I thought there was something in the past. I don't remember much about it.
I guess now we can tell them to flip their screens over and stop their bitchin'!

**Comment by AnHardt** [comment]
Fixed Turkish and Greek fonts to have decent -1. So moving the status-line down by 1 over the screenend should be save now. (Change font decent to -1 for tr and gr #5298)

**Comment by thinkyhead** [comment]
Ah, here it is: #2958

**Comment by thinkyhead** [comment]
I've reduced the hollow frame height by 1 scanline.

@esenapaj Can I assume it's only the "?" character that is hitting the top of the hollow frame with the smaller font? If so, then perhaps the text baseline can be left where it is, because the numbers are still centered.

How does it look with `XYZ_HOLLOW_FRAME` disabled?

**Comment by AnHardt** [comment]
Oh. Sorry.
From u8glib wiki:
> The reference point (0,20) for the origin of the text string usually is one pixel below the lower left edge of the first character. The height of the uppercase letters is shown in the font overview bitmap (in this example 18, see here). In some cases the size of the uppercase letters is also part of the font name.

So now the complete Gr, Tr fonts are visible at index 63 instead of cutting the lowest parts of some symbols. (if the doku is right).

@thinkyhead 
Where did you see the 2pixel distance? Only with `USE_SMALL_INFOFONT`?

**Comment by ghost** [comment]
>Can I assume it's only the "?" character that is hitting the top of the hollow frame?

Yes, it is.
Only the "?" character that is hitting the top of the hollow frame and the box, when `USE_SMALL_INFOFONT` is enabled.

| without both `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | without both `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` |
| :-: | :-: | :-: | :-: |
| `XYZ_HOLLOW_FRAME` is enabled | `XYZ_HOLLOW_FRAME` is disabled | `XYZ_HOLLOW_FRAME` is enabled | `XYZ_HOLLOW_FRAME` is disabled |
| It's displayed properly | It's displayed properly | It's displayed properly but "?" is hitting | It's displayed properly but "?" is hitting |
dscn1760 | dscn1763 | dscn1761 | dscn1762

A branch that used for testing: https://github.com/esenapaj/Marlin/tree/glcd-testes

**Comment by AnHardt** [comment]
@esenapaj 
You will not see any difference on the status-screen when `USE_BIG_EDIT_FONT` is defined. It is only used on the edit-screens. For example in Main/Control/Temperature/Nozzle: .

**Comment by ghost** [comment]
About `USE_BIG_EDIT_FONT`, I'm not sure that what is difference between when it's enabled and disabled.
It looks like that result of those are same.

| without both `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | without both `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` | `USE_BIG_EDIT_FONT` and `USE_SMALL_INFOFONT` |
| :-: | :-: | :-: | :-: |
| without `LCD_SCREEN_ROT_180` | without `LCD_SCREEN_ROT_180` | with `LCD_SCREEN_ROT_180` | with `LCD_SCREEN_ROT_180` |
dscn1766 | dscn1767 | dscn1765 | dscn1764

**Comment by AnHardt** [comment]
Seems to be already broken in RCBugFix.
An other good sign for - this is not used.

**Comment by AnHardt** [comment]
Went back to RC5 to show how it used to look like.
picture 14

**Comment by ghost** [comment]
RC6... OK
RC7... no effect

Therefore, `USE_BIG_EDIT_FONT` was broken at some point between RC6 and RC7.

https://github.com/MarlinFirmware/Marlin/commit/ecd5e810accde02b14bbb5e1611da7c05dcdf0dc ... OK
https://github.com/MarlinFirmware/Marlin/commit/9e8b3e5ee6bd7e64d05a170fb40721729106abbd ... OK
https://github.com/MarlinFirmware/Marlin/commit/1e14df876de58560e1f94fe63ab1a6b425a0dabf ... OK
https://github.com/MarlinFirmware/Marlin/commits/1e14df876de58560e1f94fe63ab1a6b425a0dabf

https://github.com/MarlinFirmware/Marlin/commit/a8b6b6a128d19178119321d2ddbe872144ef3261 ...no effect
https://github.com/MarlinFirmware/Marlin/commits/a8b6b6a128d19178119321d2ddbe872144ef3261

https://github.com/MarlinFirmware/Marlin/commit/cdd77d23bb8b195c6df37740d7c55339a64a37bd ...no effect
https://github.com/MarlinFirmware/Marlin/commit/fdea8b6e6afe7390332a8fb6bd3eb2deb928b551 ...no effect
https://github.com/MarlinFirmware/Marlin/commit/1f3283e3a56f5b5131882b787bad2704a6831f91 ...no effect
https://github.com/MarlinFirmware/Marlin/commit/2d7fe9dc7828db2370693a3bae123e81582d139f ...no effect
https://github.com/MarlinFirmware/Marlin/commit/8a2efd11550745d0816dfa7679347ed3e792368c ... no effect

Found.
~~It looks like that https://github.com/MarlinFirmware/Marlin/commit/a8b6b6a128d19178119321d2ddbe872144ef3261 broke the `USE_BIG_EDIT_FONT`, but I'm not sure that what in this commit destroyed it.
https://github.com/MarlinFirmware/Marlin/commits/1.1.0-RC7/Marlin?after=2X3BAhbZpDEngFLltglzuv7u1%2BArMzE0~~

~~Maybe It looks like that one of PRs as follows broke the `USE_BIG_EDIT_FONT`, but I'm not sure that what in those PRs destroyed it.~~

~~Replace PLA/ABS preheating with generic by thinkyhead · Pull Request #4252 · MarlinFirmware/Marlin
Don't use UTF-strlen() if the text is not UTF by AnHardt · Pull Request #4281 · MarlinFirmware/Marlin
Remove extraneous conditions for encoder position by thinkyhead · Pull Request #4264 · MarlinFirmware/Marlin~~

When I revert the PR #4281 (Don't use UTF-strlen() if the text is not UTF) from https://github.com/MarlinFirmware/Marlin/commit/a8b6b6a128d19178119321d2ddbe872144ef3261 by hand, `USE_BIG_EDIT_FONT` works properly.

**Comment by ghost** [comment]
And, I created experimental branches and tested.

| RCBugFix + revert | RCBugFix + revert | RCBugFix + Selective + revert | RCBugFix + Selective + revert |
| :-: | :-: | :-: | :-: |
| `USE_BIG_EDIT_FONT` | `USE_BIG_EDIT_FONT` | `USE_BIG_EDIT_FONT` | `USE_BIG_EDIT_FONT` |
| without `LCD_SCREEN_ROT_180` | with `LCD_SCREEN_ROT_180` | without `LCD_SCREEN_ROT_180` | with `LCD_SCREEN_ROT_180` |
| It's displayed properly | It's displayed properly | Top line of char missing | Top line of char missing |
dscn1769 | dscn1770 | dscn1768 | dscn1771

Branches that used for testing:
https://github.com/esenapaj/Marlin/tree/RCBugFix-test (RCBugFix + revert)
https://github.com/esenapaj/Marlin/tree/glcd-testes2 (RCBugFix + Selective + revert)

**Comment by olikraus** [comment]
> I've started to look at the code of U8glib and I can already see some room for optimization. Maybe this is one area where we can help U8glib itself to get better too.

I have mostly closed development for u8glib. However any help is wellcome for u8g2. 

U8g2 already includes several optimizations. For example 180 degree rotation can be done via hardware if this is supported by the display (supported by SSD1306 controller, but unfortunately not by the ST7920).
Moreover the software rotation (required for ST7920) has already been optimized.

**Comment by thinkyhead** [comment]
@esenapaj It might not be the commit you reverted that broke it, necessarily. Maybe it was broken later, because:

> "You have to define one of the DISPLAY_INPUT_CODE_MAPPERs in your language_xx.h file"
> // should not occur because (en) will set.

**Comment by thinkyhead** [comment]
@esenapaj @Blue-Marlin Any thought as to what it is in #4281 that the large font depends on?

**Comment by thinkyhead** [comment]
Maybe it's not what _was_ done in #4281 that broke `USE_BIG_EDIT_FONT`, but what was _not done_. Should these lines also have been updated?
```cpp
#if DISABLED(MAPPER_C2C3) && DISABLED(MAPPER_NON) && ENABLED(USE_BIG_EDIT_FONT)
  #undef USE_BIG_EDIT_FONT
#endif
```

**Comment by thinkyhead** [comment]
Seems like the large font is over-estimating its height. Is it really 18 pixels tall? I only count 10 pixels above the baseline. How many are below?

**Comment by ghost** [comment]
>Should these lines also have been updated?

I think that there is a possibility.
When I tried to create a experimental branch (RCBugFix + Selective + revert), [compilation error](https://github.com/esenapaj/Marlin/commit/ea5681b2a297981124f574abbfd99bdbe910894d) occurred in `#if ENABLED(USE_BIG_EDIT_FONT)` section.
It means that since PR #4281, even if user enable `USE_BIG_EDIT_FONT`, it has been disabled automatically.

**Comment by thinkyhead** [comment]
I believe I've fixed the issue of the edit text missing the top line.

We should check to see if text that contains lower-case "g" shows up, and how big the descent is.

And we should check the edit items that need 2 lines to display (because the label is long).

**Comment by thinkyhead** [comment]
> I think that there is a possibility.

So maybe that should be this instead? (Just guessing!)
```cpp
#if DISABLED(MAPPER_NON) && DISABLED(MAPPER_C2C3) && DISABLED(MAPPER_D0D1) \
  && DISABLED(MAPPER_D0D1_MOD) && DISABLED(MAPPER_E382E383) && DISABLED(MAPPER_CECF)
  #undef USE_BIG_EDIT_FONT
#endif
```

**Comment by ghost** [comment]
>So maybe that should be this instead? (Just guessing!)

It seems that it's incorrect unfortunately.
Because I'm just testing newest this PR + that solution, when I enable `USE_BIG_EDIT_FONT`, compilation error __doesn't__ occurs.
[The current code should cause an error when `USE_BIG_EDIT_FONT` is enabled.](https://github.com/MarlinFirmware/Marlin/pull/5288/files#diff-695a0b5d0447a1f65455a945b7de113dR733) Because it try to change a __const__ variable in `#if ENABLED(USE_BIG_EDIT_FONT)` section.
It means that still `USE_BIG_EDIT_FONT` has been disabled automatically.

**Comment by thinkyhead** [comment]
Unfortunately I don't know enough about #4281 or the big font to know why it is being `#undef`ined in those cases. My guess would be that it should only be set for Western languages. So there should be a more specific test for that.

And I think probably it should be caught in `SanityCheck.h` instead of just forcing it off.

**Comment by thinkyhead** [comment]
Maybe this?
```cpp
#if DISABLED(DISPLAY_CHARSET_ISO10646_1)
  #undef USE_BIG_EDIT_FONT
  #undef USE_SMALL_INFOFONT
#endif
```
That would correspond with:
```cpp
#if DISABLED(SIMULATE_ROMFONT) \
 && DISABLED(DISPLAY_CHARSET_ISO10646_1) \
 && DISABLED(DISPLAY_CHARSET_ISO10646_5) \
 && DISABLED(DISPLAY_CHARSET_ISO10646_KANA) \
 && DISABLED(DISPLAY_CHARSET_ISO10646_GREEK) \
 && DISABLED(DISPLAY_CHARSET_ISO10646_CN) \
 && DISABLED(DISPLAY_CHARSET_ISO10646_TR)
  #define DISPLAY_CHARSET_ISO10646_1 // use the better font on full graphic displays.
#endif
```
…and…
```cpp
// Western only. Not available for Cyrillic, Kana, Turkish, Greek, or Chinese.
```

**Comment by ghost** [comment]
>Maybe this?

Maybe so. At least, I've confirmed that `USE_BIG_EDIT_FONT` works properly, and "top line missing" problem is solved in English just now.
Also `LCD_SCREEN_ROT_180` is ok.

About "g", I test now.

**Comment by ghost** [comment]
How about?

| `USE_BIG_EDIT_FONT` | without `USE_BIG_EDIT_FONT` |
| :-: | :-: |
| with `LCD_SCREEN_ROT_180` | with `LCD_SCREEN_ROT_180` |
dscn1772 | dscn1773

**Comment by thinkyhead** [comment]
Looks good. I count 13 pixels tall for the big font. I'll see about getting it better-centered based on that info, and perhaps we can get one less render on 8-stripe displays too.

**Comment by ghost** [comment]
It looks like that
```cpp
#if DISABLED(DISPLAY_CHARSET_ISO10646_1)
  #undef USE_BIG_EDIT_FONT
  #undef USE_SMALL_INFOFONT
#endif
```

is forgotten.

**Comment by thinkyhead** [comment]
> It looks like that … is forgotten

I'm suggesting that you try using those lines in place of these:
```cpp
#if DISABLED(MAPPER_C2C3) && DISABLED(MAPPER_NON) && ENABLED(USE_BIG_EDIT_FONT)
  #undef USE_BIG_EDIT_FONT
#endif
```

**Comment by ghost** [comment]
Well... I used it already at test in https://github.com/MarlinFirmware/Marlin/pull/5288#issuecomment-263058556

Perhaps I may be misunderstanding the interaction with you cause by my bad English.

**Comment by Blue-Marlin** [comment]
```diff
@@ -52,12 +52,13 @@
 
 #if ENABLED(SHOW_BOOTSCREEN) && ENABLED(SHOW_CUSTOM_BOOTSCREEN)
   #include "_Bootscreen.h"
 #endif
 
-#if DISABLED(MAPPER_C2C3) && DISABLED(MAPPER_NON) && ENABLED(USE_BIG_EDIT_FONT)
+#if DISABLED(DISPLAY_CHARSET_ISO10646_1)
   #undef USE_BIG_EDIT_FONT
+  #undef USE_SMALL_INFOFONT
 #endif
 
 #if ENABLED(USE_SMALL_INFOFONT)
   #include "dogm_font_data_6x9_marlin.h"
   #define FONT_STATUSMENU_NAME u8g_font_6x9
```
is enough to bring back `USE_BIG_EDIT_FONT` to RCBugFix. Tested with en and de.
It's the most direct condition we can test. Small and big font are both coded in `ISO10646_1` so only when `ISO10646_1` is used for the 'main' font it makes sense to use the other.

**Comment by ghost** [comment]
Also, I confirmed no problem with kana and kana_utf8.

**Comment by thinkyhead** [comment]
> Perhaps I may be misunderstanding the interaction with you cause by my bad English.

From your statement "It looks like that (code) is forgotten" I thought you meant something else. The words "is forgotten" are quite poetic in this context.

I'm glad to hear that fixes it. I will add that to the PR and merge this thing. We can continue to patch this up in the coming days, but I think it's showing great promise.

**Comment by AnHardt** [comment]
I'll make a separate PR to fix 'USE_BIG_EDIT_FONT' only.

**Comment by ghost** [comment]
>The words "is forgotten" are quite poetic in this context.

Oh sorry...

**Comment by thinkyhead** [comment]
http://sourcecodepoetry.com/

image

**Comment by AnHardt** [comment]
> And we should check the edit items that need 2 lines to display (because the label is long).

Works.

**Comment by AnHardt** [comment]
About u8g2:

> Last time I checked they were still missing PROGMEM support.

the solution for this is looking like:
```
#define printP(s) print(F((s)))
#define TEST_STR "Test"

u8g2.printP(TEST_STR);
// or
u8g2.printP("An other String");
```
 
Fonts are already in PROGMEM. 
 
Something else missing? - besides of speed?

Last night's patch for u8g2 is a big step forward - but not enough. :-(
https://www.youtube.com/watch?v=c6XYxvCf2Fc

**Comment by landodragon141** [comment]
It looks like he just posted another performance update for SW SPI.
I don't think there is a video up yet.

**Comment by AnHardt** [comment]
u8g vs u8g2 - now maybe a bit too fast - somewhere 
https://www.youtube.com/watch?v=ZChlUhMJ3fQ
Downloaded u8g2 ~0:00UTC

@olikraus 
```
if ( b == 0 )
{
  *arduino_data_port &= arduino_data_n_mask;
  for( i = 0; i < 4; i++ )
  {
    *arduino_clock_port |= arduino_clock_mask;      
    *arduino_clock_port &= arduino_clock_n_mask;
          __asm__("nop\n\t");                             ///////////////////////////
    *arduino_clock_port |= arduino_clock_mask;      
    *arduino_clock_port &= arduino_clock_n_mask;
  }
}
else
```
A nop in both of the 0-loops seems to be enough to have no pixel errors on my display.

**Pull request merged**